src/OpenTelemetry/Trace/TracerFactory.cs

```diff
@@ -0,0 +1,77 @@
+﻿// <copyright file="TracerFactory.cs" company="OpenTelemetry Authors">
+// Copyright 2018, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+namespace OpenTelemetry.Trace
+{
+    using System.Collections.Generic;
+    using OpenTelemetry.Context.Propagation;
+    using OpenTelemetry.Resources;
+    using OpenTelemetry.Trace.Config;
+    using OpenTelemetry.Trace.Export;
+
+    /// <inheritdoc/>
+    public sealed class TracerFactory : ITracerFactory
+    {
+        private readonly object lck = new object();
+        private readonly SpanProcessor spanProcessor;
+        private readonly TraceConfig traceConfig;
+        private readonly Dictionary<string, ITracer> tracerRegistry = new Dictionary<string, ITracer>();
+
+        public TracerFactory(SpanProcessor spanProcessor = null, TraceConfig traceConfig = null)
+        {
+            this.spanProcessor = spanProcessor ?? Tracing.SpanProcessor;
+            this.traceConfig = traceConfig ?? Tracing.TraceConfig;
+            this.TextFormat = new TraceContextFormat();
+            this.BinaryFormat = new BinaryFormat();
+        }
+
+        internal ITextFormat TextFormat { get; set; }
+        
+        internal IBinaryFormat BinaryFormat { get; set; }
+
+        /// <inheritdoc/>
+        public ITracer GetTracer(string name, string version = null)
+        {
+            lock (this.lck)
```

Comment:
my guess is that in 99% of cases there will be only one tracer - no registry is needed. Can we separate default tracer from this dict and if name if null or empty simply return it without locking?

Reply:
This very much against the whole idea of having named tracers and meters and the name is actually a required parameter. Returning a default tracer for the empty name should merely be seen as a fallback.

However, we could still handle that case separately.